File: 프로그래머스/코딩테스트/게임 맵 최단거리.cpp
#include <vector>
#include <algorithm>
#include <queue>
#include<iostream>
using namespace std;

int visited[101][101];
int dx[] = { -1,0,1,0 };
int dy[] = { 0,-1,0,1 };
int n, m;

queue<pair<int, int>> q;

void bfs(vector<vector<int> > maps, int x, int y) {
	q.push({ x,y });
	visited[x][y] = 1;

	while (!q.empty()) {
		int cx = q.front().first;
		int cy = q.front().second;
		q.pop();

		for (int i = 0; i < 4; i++) {
			int nx = cx + dx[i];
			int ny = cy + dy[i];
			if (nx < 0 || ny < 0 || nx >= n || ny >= m)continue;
			if (!visited[nx][ny] && maps[nx][ny] == 1) {
			
				if (visited[nx][ny] > visited[cx][cy] + 1) {
					visited[nx][ny] = visited[cx][cy] + 1
					q.push({ nx, ny });
				}
			}
		}
	}
}
int solution(vector<vector<int> > maps)
{
	n = maps.size();
	m = maps[0].size();

	bfs(maps, 0, 0);
	int answer = 0;

	answer = visited[n - 1][m - 1];
	if (answer == 0) {
		answer = -1;
	}

    return answer;
}

File: 프로그래머스/코딩테스트/게임 맵 최단거리.py
from collections import deque
visited = [[0]*101 for _ in range(101)]
dx = [-1, 0, 1, 0]
dy = [0, -1, 0, 1]

def bfs(maps):
    q = deque()
    q.append([0,0])
    while q:
        cx, cy = q.popleft()
        for i in range(4):
            nx = cx + dx[i]
            ny = cy + dy[i]
            if nx<0 or ny < 0  or nx >=len(maps) or ny >=len(maps[0]):
                continue
            if not visited[nx][ny] and maps[nx][ny] == 1:
                visited[nx][ny] = 1
                maps[nx][ny] = maps[cx][cy] + 1
                q.append([nx, ny])
                
def solution(maps):
    answer = 0
    bfs(maps)
    if maps[-1][-1] == 1:
        answer = -1
    else:
        answer = (maps[-1][-1])
        
    return answer
